[Feature/sw 23] 메인 모듈 추가 및 화면 이동과 API 호출 세팅

SeonJeongk opened the pull request:
## 작업 내용

> 메인 모듈 추가 및 화면 이동과 API 호출 세팅

- **main 모듈 추가 및 NavHost 세팅**
    - feature 내부에 main 모듈 추가
    - MainActivity 이동 ➡️ `themes.xml`은 app에서 core 모듈로 이동
    - MainScreen에 **NavHost**구현
        - HomeScreen의 버튼을 누르면 DetailScreen으로 이동
        - 이때, 누른 버튼의 정보(`노선`)를 함께 넘겨 detail 화면에 부여주도록 설정
            - 🔔 test용 버튼이므로 리스트로 바꾸면서 삭제할 예정
            - ex) `1호선 역 정보로 이동` 버튼 클릭
                <img src="https://github.com/user-attachments/assets/d67c75ab-f79c-4c49-82bb-4312a320d1f5" width="220"/> <img src="https://github.com/user-attachments/assets/ba503990-130f-45b7-a650-fb9e8addd941" width="220"/>

- **NetworkModule 수정**
    - `ApiKeyInterceptor` encodedPath 방식으로 수정 (`format: BASE_URL/{API_KEY}/json`)
    - JSON Converter 모듈화
    - `SerializationModule` 삭제하고 _NetworkModul로 통합_

- **기본 화면 세팅**
    - **Home**과 **Detail**에 `Screen`, `Contract`, `ViewModel` 생성
    - ui 구현하면서 수정할 예정
    
- **API 세팅**
    - Response들 구현
    - SubwayService 구현
        - `getAllLineInfo` : [지하철 전 노선 및 노선별 역 정보 가져오기](https://data.seoul.go.kr/dataList/OA-15442/S/1/datasetView.do) ➡️ Home
        - `getStationInfo` : [지하철 노선 별 역별 승하차 인원 가져오기](https://data.seoul.go.kr/dataList/OA-12914/S/1/datasetView.do) ➡️ Detail
    - UseCase 구현
    - MainRepository
        - 데이터 엑세스 메서드 구현
    - MainRepositoryImpl
        - runCatching을 활용해 API 호출 결과 처리

- **각종 의존성 추가 및 수정**
    - 플러그인 적용 (hilt, kotlin serialization)



## 실행 화면

- **실행 화면**

    https://github.com/user-attachments/assets/85449bc8-43d7-4626-9e35-c2a1885280d9

- **API 호출 테스트 결과**
![모든 역](https://github.com/user-attachments/assets/137a8377-315f-4dd3-94a6-7573c92c59ea)
![모든 노선](https://github.com/user-attachments/assets/c40134c4-900e-4718-8b1d-3cc1decbcb37)


## 리뷰 요청 사항
500줄 넘어갔습니다..🧨 죄송함미다


## 레퍼런스
https://data.seoul.go.kr/dataList/OA-15442/S/1/datasetView.do
https://data.seoul.go.kr/dataList/OA-12914/S/1/datasetView.do

SeonJeongk committed:
[SW-31] chore: build.gradle 수정

SeonJeongk committed:
[SW-32] move: themes 위치 core로 이동

SeonJeongk committed:
[SW-32] feat: main 모듈 생성 및 MainActivity 이동

SeonJeongk committed:
[SW-31] chore: plugin 수정

SeonJeongk committed:
[SW-28] fix: 네트워크 모듈 수정

SeonJeongk committed:
[SW-23] feat: home, detail 모듈 화면 세팅

SeonJeongk committed:
[SW-32] feat: NavHost 기본 세팅

SeonJeongk committed:
[SW-32] feat: Home에서 Detail 화면으로 노선 정보 넘겨주기

SeonJeongk committed:
[SW-33] feat: SubwayService 및 Response 생성

SeonJeongk committed:
[SW-34] feat: MainRepository, MainRepositoryImpl 메서드 구현

SeonJeongk committed:
[SW-16] remove: 파일 삭제

SeonJeongk committed:
[SW-33] fix: API 호출 수정

SeonJeongk committed:
[SW-33] move: 파일 위치 수정

SeonJeongk committed:
[SW-16] chore: Application class명 및 gradle 수정

- domain에 inject 의존성, serialization 플러그인 추가

SeonJeongk committed:
[SW-34] fix: Response 수정

- @Serializable 추가
- 상위 키 변경(SearchSTNBySubwayLineInfo, CardSubwayStatsNew)

SeonJeongk committed:
[SW-34] feat: MainRepositoryImpl 수정 및 SubwayService 업데이트

- runCatching을 활용한 API 호출 결과 처리 로직 추가
- getStationInfo 메서드에 date와 line 매개변수 추가
- SubwayService Result<> 제거

SeonJeongk committed:
[SW-34] feat: UseCase 구현

SeonJeongk committed:
[SW-34] fix: 범위 수정 및 ViewModel에 UseCase 추가

hoyahozz commented on data/src/main/java/com/example/data/repository/MainRepositoryImpl.kt:
```diff
@@ -1,8 +1,38 @@
+    private val subwayService: SubwayService,
+) : MainRepository {
+    override suspend fun getAllLineInfo(): Result<LineInfoResponse> {
+        return runCatching {
+            subwayService.getAllLineInfo()
+        }.fold(
```
`fold` 없어도 `Result` 타입으로 리턴돼서, 하단 로직은 필요하지 않을 것 같아요 !

hoyahozz commented on domain/src/main/java/com/example/domain/model/BaseResponse.kt:
```diff
@@ -0,0 +1,9 @@
+
+@Serializable
+data class Result (
+    val CODE: String,
+    val MESSAGE: String
```
코틀린 컨벤션에 따라 소문자로 변경하고, `SerialName` 붙여주는게 좋을 것 같아요 ~!

hoyahozz commented on domain/src/main/java/com/example/domain/model/LineInfoResponse.kt:
```diff
@@ -0,0 +1,27 @@
+package com.example.domain.model
```
모든 `Response` 모델 `data` 모듈로 이관하고, `domain` 계층 모델 생성하기 !

SeonJeongk replied:
**도전 !**

SeonJeongk committed:
[SW-23] refactor: sideEffect data object로 수정

hoyahozz commented on feature/detail/src/main/java/com/example/detail/DetailContract.kt:
```diff
@@ -0,0 +1,16 @@
+    data object DetailUiEvent : UiEvent
+
+    sealed class DetailSideEffect : UiSideEffect {
+        object ShowToast : DetailSideEffect()
```
`data object` 로 변경하기 !

hoyahozz commented on feature/detail/src/main/java/com/example/detail/DetailScreen.kt:
```diff
@@ -0,0 +1,48 @@
+
+@Composable
+fun DetailScreen(
+    modifier: Modifier = Modifier,
```
`Screen` 에선 `Modifier` 매개변수로 받지 않도록 변경하기 !

hoyahozz reviewed (changes requested):
구두로 말씀드렸던 내용 리뷰로 남겨놓았습니다 ~!

화이팅 ! 👍👍👍

SeonJeongk committed:
[SW-23] refactor: Screen modifier 매개변수 제거

SeonJeongk committed:
[SW-34] refactor: fold 삭제

- 성공/실패의 추가적인 처리 로직이 필요 없다

SeonJeongk committed:
[SW-34] refactor: BaseResponse에 SerialName 추가

SeonJeongk committed:
[SW-34] feat: Response 위치 이동 및 DataSource 도입

- Response model들 domain->data 이전
- Response toDomain() 구현 -> domain 계층 모델로 변환
- DataSource 구현
- DataSource에 맞게 기존 코드 수정

SeonJeongk commented:
> 구두로 말씀드렸던 내용 리뷰로 남겨놓았습니다 ~!
> 
> 화이팅 ! 👍👍👍

리뷰 내용 모두 반영했습니다!

hoyahozz commented on data/src/main/java/com/ssun/subway/data/source/SubwayDataSource.kt:
```diff
@@ -0,0 +1,18 @@
+    private val subwayService: SubwayService,
+) {
+    suspend fun getAllLineInfo(): LineInfoResponse {
+        return subwayService.getAllLineInfo()
```
'데이터 소스에서 데이터를 불러오는 과정에서 오류가 발생할 수 있다' 라는 개념이기 때문에, `runCatching` 은 `dataSource` 단에서 붙여주는게 좋을 것 같아요!

hoyahozz commented on domain/src/main/java/com/ssun/subway/domain/model/DetailStationInfo.kt:
```diff
@@ -0,0 +1,9 @@
+package com.example.domain.model
+
+data class DetailStationInfo(
```
어떤 것의 `Detail` 일까용? 명확하게 `Station` 이라는 이름은 어떨까요?

SeonJeongk replied:
Detail 화면에서 쓰여서 Detail을 붙였습니다! `Station`이 더 명확할 것 같네용

hoyahozz replied:
도메인 모델을 설계할 땐, 어제 다같이 이야기했었던 정책 정의서를 기반으로 설계하는 것이 가장 좋아요!

그렇게 설계하면, 정책이 변경되지 않는 한 도메인 모듈에는 변경이 일어나지 않게 됩니다!

사용자가 필요로 하는 기능이 무엇이고, 이에 따라 앱에서 어떤 데이터를 표현해야 할지 생각해보면 좋을 것 같아요 :)

hoyahozz commented on domain/src/main/java/com/ssun/subway/domain/model/HomeLineInfo.kt:
```diff
@@ -0,0 +1,9 @@
+package com.example.domain.model
+
+data class HomeLineInfo(
```
`Info` 라는 접미사를 꼭 붙여야 하는 이유가 있을까여 ?.?

SeonJeongk replied:
없네여.. 지우겠습니다!

hoyahozz commented on domain/src/main/java/com/ssun/subway/domain/model/DetailStationInfo.kt:
```diff
@@ -0,0 +1,9 @@
+package com.example.domain.model
+
+data class DetailStationInfo(
+    val usageDate: String,
```
`LocalDateTime` 으로 데이터 타입을 명확히 하면 좋을 것 같아요!

hoyahozz commented on data/src/main/java/com/ssun/subway/data/model/LineInfoResponse.kt:
```diff
@@ -25,3 +26,15 @@ data class StationInfo(
     @SerialName("STATION_NM_JPN") val stationNameJpn: String,
 )
+
+fun LineInfoResponse.toDomain(): List<HomeLineInfo> {
```
`HomeLineInfo` 는 `StationInfo` 만으로도 충분히 생성할 수 있지 않을까여?

`LineInfoResponse` 는 `StationInfo` 에 비해 좀 더 넓은 개념같아요 '-'

같은 추상화 레벨끼리 대응하게 하는 것이, 추후에 변경사항이 생겼을 때도 대응이 쉽습니다!

hoyahozz commented on data/src/main/java/com/ssun/subway/data/model/StationDetailResponse.kt:
```diff
@@ -24,3 +25,15 @@ data class StationsInfo(
+
+fun StationDetailResponse.toDomain(): List<DetailStationInfo> {
+    return searchResult.row.map { result ->
+        DetailStationInfo(
```
여기도 마찬가지로, 추상화 레벨을 동일하게 맞춰주시고, `map` 은 `Repository` 에서 실행하도록 만드는 것이 좀 더 자연스러워 보여요!

`DataSource`, `Repository`, `Usecase` 의 역할은 무엇일까요?

hoyahozz commented on domain/src/main/java/com/ssun/subway/domain/repository/MainRepository.kt:
```diff
@@ -1,5 +1,9 @@
 
+interface MainRepository {
+    suspend fun getAllLineInfo(): Result<List<HomeLineInfo>>
+    suspend fun getStationInfo(date: String, line: String): Result<List<DetailStationInfo>>
```
```suggestion
    suspend fun getStationInfo(
       date: String, 
       line: String,
    ): Result<List<DetailStationInfo>>
```

트레일링 콤마 넣어주세요 ~~!

hoyahozz reviewed (changes requested):
추가적으로 리뷰 더 남겨보았습니당 ㅎㅎ

코멘트에 대해서 한 번 생각해보시고, 답변 남겨주세요 ~~!

SeonJeongk committed:
[SW-34] feat: 홈화면 레이아웃 변경 및 문구 수정

SeonJeongk committed:
[SW-34] fix: 승하차 통계 응답 타입을 문자열 기준으로 보정

SeonJeongk committed:
[SW-34] feat: 상세 화면에 호선별 역 승하차 통계 표시 추가

SeonJeongk committed:
[SW-34] feat: 홈 화면에 API 기반 호선 목록 불러오기

SeonJeongk committed:
[SW-34] feat: 승하차 통계 예외 처리 추가

SeonJeongk committed:
[SW-17] chore: 패키지명 변경

SeonJeongk committed:
[SW-17] chore: 앱 아이콘 적용

SeonJeongk committed:
[SW-00] chore: 버전코드 업데이트 (1->2)
